SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp

```diff
@@ -2546,18 +2547,31 @@ void USpatialNetDriver::TryFinishStartup()
 {
 	if (IsServer())
 	{
+		// Limit Log frequency.
+		bool bShouldLogStartup = [this]() {
+			const uint64 WatchdogTimer = 5.0 / FPlatformTime::GetSecondsPerCycle64();
+			const uint64 CurrentTime = FPlatformTime::Cycles64();
+			if (CurrentTime - StartupTimestamp > WatchdogTimer)
+			{
+				StartupTimestamp = CurrentTime;
+				return true;
+			}
```

Comment:
Have you tested this code? Your StartupTimestamp and StartupDebugString variables seem to be set on the client/in client delegate functions, then attempted to be updated or printed on the server? 

To be explicit, please can you check the debuggability / spammy nature of our startup output:
- in server logs when too few servers are launched for the lbs strat.
- in server logs when too many servers are launched for the lbs strat.
- in client logs when either of the above cases are true.
- in client logs when the client connects before the servers.

Reply:
? Did I test the code ? Well yes, what a strange question. I expect that the delegates you are calling "client delegates" are on servers which are not authoritative over the GSM.
This code has a few issues, as pointed out in a comment above, namely that the startups log should be gathered in one place if we want to control their spammy nature, and that would be a bigger undertaking, with maybe a bad ROI right since we will want to change a lot of the startup with USLB.

When not enough server connects : 

On server with auth over the VirtualWorkerManager (very spammy) :

[2020.09.29-14.40.29:950][365]LogSpatialVirtualWorkerTranslationManager: Waiting for all virtual workers to be assigned before publishing translation update. We currently have 2 workers connected out of the 4 required
[2020.09.29-14.40.29:950][365]LogSpatialVirtualWorkerTranslator: Received invalid mapping, server is still waiting for a valid mapping from the virtual worker translation manager.
[2020.09.29-14.40.29:983][366]LogSpatialVirtualWorkerTranslationManager: Waiting for all virtual workers to be assigned before publishing translation update. We currently have 2 workers connected out of the 4 required
[2020.09.29-14.40.29:983][366]LogSpatialVirtualWorkerTranslator: Received invalid mapping, server is still waiting for a valid mapping from the virtual worker translation manager.
[2020.09.29-14.40.30:016][367]LogSpatialVirtualWorkerTranslationManager: Waiting for all virtual workers to be assigned before publishing translation update. We currently have 2 workers connected out of the 4 required
[2020.09.29-14.40.30:016][367]LogSpatialVirtualWorkerTranslator: Received invalid mapping, server is still waiting for a valid mapping from the virtual worker translation manager.
[2020.09.29-14.40.30:049][368]LogSpatialVirtualWorkerTranslationManager: Waiting for all virtual workers to be assigned before publishing translation update. We currently have 2 workers connected out of the 4 required
[2020.09.29-14.40.30:049][368]LogSpatialVirtualWorkerTranslator: Received invalid mapping, server is still waiting for a valid mapping from the virtual worker translation manager.
[2020.09.29-14.40.30:082][369]LogSpatialVirtualWorkerTranslationManager: Waiting for all virtual workers to be assigned before publishing translation update. We currently have 2 workers connected out of the 4 required
[2020.09.29-14.40.30:082][369]LogSpatialVirtualWorkerTranslator: Received invalid mapping, server is still waiting for a valid mapping from the virtual worker translation manager.
[2020.09.29-14.40.30:115][370]LogSpatialVirtualWorkerTranslationManager: Waiting for all virtual workers to be assigned before publishing translation update. We currently have 2 workers connected out of the 4 required
[2020.09.29-14.40.30:115][370]LogSpatialVirtualWorkerTranslator: Received invalid mapping, server is still waiting for a valid mapping from the virtual worker translation manager.
[2020.09.29-14.40.30:149][371]LogSpatialVirtualWorkerTranslationManager: Waiting for all virtual workers to be assigned before publishing translation update. We currently have 2 workers connected out of the 4 required
[2020.09.29-14.40.30:149][371]LogSpatialVirtualWorkerTranslator: Received invalid mapping, server is still waiting for a valid mapping from the virtual worker translation manager.

On other servers (every 5 sec):

[2020.09.29-14.39.49:826][153]LogSpatialOSNetDriver: Waiting for the GSM to be ready : waiting for query
[2020.09.29-14.39.54:837][305]LogSpatialOSNetDriver: Waiting for the GSM to be ready : waiting for query
[2020.09.29-14.39.59:848][457]LogSpatialOSNetDriver: Waiting for the GSM to be ready : waiting for query
[2020.09.29-14.40.04:856][609]LogSpatialOSNetDriver: Waiting for the GSM to be ready : waiting for query
[2020.09.29-14.40.09:861][761]LogSpatialOSNetDriver: Waiting for the GSM to be ready : waiting for query
[2020.09.29-14.40.14:872][913]LogSpatialOSNetDriver: Waiting for the GSM to be ready : waiting for query
[2020.09.29-14.40.19:879][ 65]LogSpatialOSNetDriver: Waiting for the GSM to be ready : waiting for query
[2020.09.29-14.40.24:890][217]LogSpatialOSNetDriver: Waiting for the GSM to be ready : waiting for query
[2020.09.29-14.40.29:904][369]LogSpatialOSNetDriver: Waiting for the GSM to be ready : waiting for query

I expect the too many server case to pass startup, and have the remaining servers be stuck on the log above.

IIRC, the client does not spam much in either case, but I do not have much detail.

Reply:
Apologies, that was poorly phrased, I mean that on servers it will only ever print `waiting for query` because `StartupDebugString` seems to only be set in `USpatialNetDriver::GSMQueryDelegateFunction` which is only called on the client? So effectively `StartupDebugString` is being set on the client but never printed, and never set on the server.

For the translation manager, can we find a way to:

1.  Ignore the Translator logs (since this is the TranslationManager workers we don't really care about the Translator log while the TranslationManager is failing)

1. Reduce how much it's spamming?

Reply:
Oh, is the GSM query not ran from servers ? I did not realize that, I'll have to dive deeper in the startup sequence.

So about your two points, that's where I say we could find a way, but that will be tying together systems that are currently separated. Either we break encapsulation, or we add some log overhaul in order to gather them at the same place to control log rate (and which logs actually matter). And that's where I am saying that I am reluctant to make an investment.

Reply:
Fixed client logs.